fpdf/text_region.py

```diff
@@ -0,0 +1,275 @@
+            raise FPDFException(
+                f"Unable to enter the same {self.__class__.__name__} context recursively."
+            )
+        self._page = self.pdf.page
+        self.pdf._push_local_stack()
+        self.pdf.page = 0
+        self.pdf.register_text_region(self)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.pdf.clear_text_region()
+        self.pdf.page = self._page
+        self.pdf._pop_local_stack()
+        self.render()
+
+    def write(self, text: str):  # , link: str = ""):
+        if self._has_paragraph == "EXPLICIT":
+            raise FPDFException(
+                "Conflicts with active paragraph. Consider adding your text there."
+            )
+        if self._has_paragraph is None:
+            p = Paragraph(region=self, align=self.align)
+            self._paragraphs.append(p)
+            self._has_paragraph = "AUTO"
+        self._paragraphs[-1].write(text)
+
+    def paragraph(self, align=None):
+        if self._has_paragraph == "EXPLICIT":
+            raise FPDFException("Unable to nest paragraphs.")
+        p = Paragraph(region=self, align=align or self.align)
+        self._paragraphs.append(p)
+        self._has_paragraph = "EXPLICIT"
+        return p
+
+    def end_paragraph(self):
+        if not self._has_paragraph:
+            raise FPDFException("No active paragraph to end.")
+        self._has_paragraph = None
+
+
+class TextRegion(ParagraphCollectorMixin):
+    """Abstract base class for all text region subclasses."""
+
+    def _ln(self, h=None):
+        self.pdf.ln(h)
+
+    def current_x_extents(
+        self, y, height
+    ):  # xpylint: disable=no-self-use,unused-argument
+        """Return the horizontal extents of the current line."""
+        raise NotImplementedError()
+
+    def _render_lines(self, text_lines):
+        page_break_triggered = False
+        self.pdf.y = max(self.pdf.y, self.pdf.t_margin)
+        text_line = None
+        for text_line_index, text_line in enumerate(text_lines):
+            if text_line_index != 0:
+                self._ln()
+            # print(self.pdf.y + text_line.height, self.pdf.page_break_trigger)
+            if hasattr(self, "accept_page_break"):
+                if self.pdf.y + text_line.height > self.pdf.page_break_trigger:
+                    page_break_triggered = self.accept_page_break()
+            new_page = self.pdf._render_styled_text_line(
+                text_line,
+                text_line.max_width,
+                h=text_line.height,
+                border=0,
+                new_x=XPos.WCONT,
+                new_y=YPos.TOP,
+                fill=False,
+                #                link=link,
+            )
+            page_break_triggered = page_break_triggered or new_page
+        if text_line and text_line.trailing_nl:
+            # The line renderer can't handle trailing newlines in the text.
+            self.pdf._ln()
```

Comment:
Based on CodeCov, this line is not covered by the tests

Reply:
There are quite a few of those. It's still a draft, after all...